docs/configuration.md
---
title: Configuration
---

# Configuration

airtrack is configured via command line flags and a configuration file. The command line
flags define some constant system wide options (location of config files, log level),
and the configuration file sets out database connection settings, email settings, project
settings, and so on.

To view the command line options run `./airtrack track --help`

## Configuration files

Airtrack uses [YAML format](https://en.wikipedia.org/wiki/YAML) for its configuration files.

Placeholders for certain types of data: 
 * `<boolean>`: a boolean taking `true` or `false` as a value
 * `<string>`: a string of characters
 * `<host>`: a valid hostname or IP address
 * `<int>`: an integer value
 * `<secret>`: a secret sequence of characters. passwords/apikeys etc
 * `<ip_address>`: an IP address
 * `<http_url>`: a valid URL with an HTTP or HTTPS schema
 * `<dirpath>`: a valid path to a directory on the filesystem (absolute or relative)
 * `<email>`: a valid email address 
 * `<filter>`: a valid CEL expression. See [Project Filters](project-filter.html) for more information.
 * `<notificationevent>`: a notification event supported by airtrack
 * `<feature>`: a tracking feature supported by airtrack
 * `<mapservice>`: a valid map layout supported by airtrack

This document contains overviews of various YML configuration blocks, the types of
each parameter, and the structures representing entire configuration files.

### Main configuration file

The `--config` command line option specifies the main configuration file.
 
The main configuration file has the following structure:
```yaml
# Database configuration
database: <database_config>

# Local time zone
[ timezone: <string> | default = system timezone ]

# Configuration for ADSBExchange API
[ adsbx: <adsbexchange_config> | default = none ]

# Configuration for beast servers
beast:
  [ - <beast_config> | default = none ]

# Import airport locations for flight source + destination geolocation
[ airports: <airports_config> | default = none ]

# Configuration for the email driver
[ email: <email_config> | default = none ]

# Configuration for the HTTP server with maps
[ map: <map_config> | default = none ]

# Configuration for prometheus metrics
[ metrics: <metrics_config> | default = none ]

# Configure system-wide defaults for all projects
[ sightings: <sightings_config> | default = none ]

# Project configurations
projects:
[ - <project_config> | default = none ]
```

### Projects configuration file
It's also possible to pass additional _projects only_ configuration files using the 
`--projects` command line option. The option can be repeated to pass several project
only configuration files.

A project configuration file has the following structure:
```yaml
# Project configurations
projects:
[ - <project_config> | default = none ]
```

### `<adsbexchange_config>`
[ADS-B Exchange](https://adsbexchange.com) is an unfiltered map of data received
by a community of ADS-B receivers around the world. They provide an API which is
free for feeders or can access can be purchased. 

An API key is required for this configuration.

```yaml
# Your API key for connection to ADS-B Exchange
apikey: <secret>
# URL for ADS-B Exchange. Not required unless you have
# a local cache of the API
[ url: <http_url> | default = "https://adsbexchange.com/api/aircraft/json/" ]
```

### `<beast_config>`
BEAST format messages are produced by dump1090 on port 30005 by default.

```yaml
# Name for this data source
name: <string>
# Hostname or IP for server
host: <host>
# Port for connection (probably 30005). Optional, as defaults
# to 30005
[ port: <port> | default = 30005 ]
```

### `<airports_config>`

Airtrack can geolocate the takeoff and landing airport for a flight.
Locations are provided to airtrack via `aip` and `cup` files. Airtrack
releases contain the latest available version of the openaip database.

Use of the built-in openaip locations can be disabled in this configuration section.
It's also possible to configure custom directories containing `aip` and `cup` files.

[Click here for information about airport location files](airport-locations.html)

```yaml
# disable_builtin_airports will prevent airtrack from loading
# compiled-in airport files if set to `true`
[ disable_builtin_airports = <boolean> | default = false ]

# List of directories containing aip files:
openaip: 
  [ - <dirpath> | default = none ]
# List of directories containing cup files:
cup:
  [ - <dirpath> | default = none ]
```

### `<email_config>`

The `<email_config>` section contains configuration related to sending email.

The `driver` field is required along with the configuration for that particular driver.
Currently, only the `smtp` driver is supported.

```yaml
# Email driver. Currently only 'smtp' is supported
driver: <string>

# Configuration for 'smtp' email driver
[ smtp: <smtp_config> | default = none ]
```

### `<smtp_config>`

The `<smtp_config>` section configures the SMTP based email driver.

The `sender`, `username`, `password`, `host`, and `port` fields are required.

To use TLS, `tls` must be set to `true`.

For connections without TLS, the default connection strategy will use STARTTLS if the
server advertises it. STARTTLS can be made mandatory by setting `mandatory_starttls`
to `true`. To disable opportunistic STARTTLS, set `nostarttls` to `true`.
 
```yaml
# sender email address
sender: <email>
# SMTP username
username: <string>
# SMTP password
password: <secret>
# The SMTP server's hostname/ip
host: <host>
# The SMTP server port
port: <int>

# Whether to connect using TLS (port 587)
[ tls: <boolean> | default = false ]

# If `mandatory_starttls` set to true, connections 
# to servers which do not advertise STARTTLS support
# will cause an error.
[ mandatory_starttls: <boolean> | default = true ] 
# disables opportunistic encryption of connection with STARTTLS
# if this is set to true
[ nostarttls: <boolean> | default = false ]
```

### `<map_config>`

The `<map_config>` section contains configuration for the map web server.

This configuration section can be omitted, and the server will start with 
the following defaults:
  - the map is accessible via any interface on port 8080
  - history files created every 30 seconds, and 60 history files are kept (30 minutes of history)
  - tar1090 and dump1090 maps available at ./dump1090/$project and ./tar1090/$project respectively.

```yaml
# Interface the HTTP server will listen on
[ interface: <ip_address> | default = "0.0.0.0" ]
# Port for the HTTP server. Default is 8080
[ port: <int> | default = 8080 ]
# Disable map server. Default is false.
[ disabled: <boolean> | default = false ]
# Number of seconds before creating a new history file. Default
# is every 30 seconds.
[ history_interval: <int> | default = 30 ]
# Number of history files to maintain. Default is to
# store 60 files. Defaults should store 30 minutes of history files.
[ history_count: <int> | default = 60 ]
# The map frontends to provide. Essentially different map skins.
services:
[ - <mapservice> | default = "tar1090", "dump1090" ]
```

### `<metrics_config>`

The `<metrics_config>` section contains confirmation for prometheus metrics about the golang internals and airtrack operations.

If the section is missing or `enabled` is false the metrics server will be disabled.

```yaml
# Whether to enable metrics server.
[ enabled: <boolean> | default = false ]
# Interface the metrics HTTP server will listen on.
[ interface: <ip_address> | default = "0.0.0.0" ]
# Port the metrics HTTP server will listen on.
[ port: <int> | default = 9206 ]
```

### `<database_config>`

A `<database_config>` section is required for airtrack to run. The supported engines are:
 * [MySQL](#mysql)
 * [PostgreSQL](#postresql)
 * [SQLite Version 3](#sqlite)  
 
```yaml
# Database driver. Currently supports 'mysql' and 'sqlite3'
driver: <string>
# DB Host - server to connect to if using mysql
host: <host>
# DB Port - server port if using mysql
port: <int>
# DB Username - username for mysql connection
username: <string>
# DB Password - password for mysql connection
password: <secret>
# Selected Database
# if driver is mysql, this is the database on the mysql server
# if driver is sqlite3, this is the filesystem path to the DB
database: <string|filepath>
```

#### SQLite

SQLite is an embedded storage engine that stores its data in a file.

The `database` option defines the location of the database on the filesystem.
The path may be relative or absolute.

A config file extract of SQLite configuration:
```yaml
# ...
driver: sqlite3
database: /home/user/airtrack.sqlite3
# ...
```

#### MySQL

MySQL is a database server that is connected to over the network.  

A config file extract of MySQL configuration:
```yaml
# ...
driver: mysql
host: server.local
port: 3306
username: airtrack
password: password
database: airtrack
# ...
```

#### PostreSQL

PostgreSQL is a database server that is connected to over the network.  

A config file extract of PostgreSQL configuration:
```yaml
# ...
driver: postgres
host: server.local
port: 3306
username: airtrack
password: password
database: airtrack
# ...
```

### `<project_config>`

A project contains information about how aircraft should be tracked, and optionally,
filters to focus the project on certain aircraft.

The projects [`map` field](#project_map_config) contains per-project configuration of 
the map display. Currently, this only allows the project to be opted out from the map.

Projects can enable `<feature>`'s to determine which data about the flight should be
tracked. [Click here for documentation of available features](project-features.html)

A project can filter aircraft from it's tracking functions if a filter is configured.
[Click here for information about project filters](project-filter.html)

If you'd like to reopen an aircraft sighting if it's visible again within
`reopen_sightings_interval` seconds, `reopen_sightings` can be set to `true`.

Sometimes due to reception problems a message may incorrectly indicate an aircraft
is on the ground. The `onground_update_threshold` defines the number of consecutive
messages to wait before accepting a new `onground` value.

```yaml
# A name for the project
name: <label>

# List of features enabled for the project
features:
[ - <feature> | default = none ]

# Configuration of the map for this project
[ map: <project_map_config> | default = none ]

# Disabled can used to prevent a project from running
[ disabled: <boolean> | default = false ]

# Set a filter for this project. Can be omitted if not required.
[ filter: <filter> | default = none ]

# Project specific notification configuration
[ notifications: <notification_config> | default = none ]

# Whether to reopen a sighting if it was closed recently, due
# to bad coverage, etc
[ reopen_sightings: <boolean> | default = false ]

# How long after an aircraft goes out of range
# before we no longer reopen a recently closed session.
# Unit: seconds.
[ reopen_sightings_interval: <int> | default = 5m ]

# Sometimes due to reception problems a message may incorrectly indicate an aircraft
# is on the ground. The `onground_update_threshold` defines the number of consecutive
# messages to wait before accepting a new `onground` value.
[ onground_update_threshold: <int> | default = 6 ]

# Set a minimum interval in seconds between accepting locations. When this is zero or
# not provided, the system-wide default is used (configured in <sightings_config>)
# Unit: seconds
[ location_update_interval: <int> | default = 0 ]
```

### `<notification_config>`

A projects `<notification_config>` section defines a list of events 
which should result in an email notification, and an email address
to send notifications to.

[Click here for documentation of available email notifications](project-event-notifications.html)
```yaml
# Destination for notifications
email: <email>

# List of enabled email notifications
events:
[ - <notificationevent> | default = none ]
```

### `<project_map_config>`

A projects `<project_map_config>` section contains per-project configuration
related to the web based map interface. 

Currently the only available option is to opt the project out of the map.

```yaml
# Disabled can be used to prevent a projects map being visible on the map.
[ disabled: <boolean> | default = false ]
```

### `<sightings_config>`

The `<sightings_config>` block allows sighting related defaults to be modified.

```yaml
# Control the sighting timeout (how long after the last message until the
# sighting will be closed). A project can also configure a custom timeout.
# Unit: seconds
[ timeout: <int> | default = 60s ]

# Control how often location updates are saved. Defaults to zero so all
# locations are recoded. If non-zero, this number of seconds must pass
# before another location will be recorded. A project can also configure
# a custom location_update_interval.
[ location_update_interval: <int> | default = 0s ]
```


docs/changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.1] - 2020-10-09

### Added

 - Adds a changelog to the project.

## [0.0.2] - 2020-10-11

### Added

 - Now automatically building releases for linux-amd64

## [0.0.3] - 2020-10-11

### Added

 - Include openAIP data with distributed releases
 - Add `version` command to airtrack

## [0.0.4] - 2020-12-15

### Added

 - Adds `location_update_interval` option to project configuration. Enables frequency
   of location updates to be managed to one per interval
 - Add `location_update_interval` option to global sightings configuration. Allows
   control of the default `location_update_interval` if a project has none configured.
 - Adds a `Signal` proto message, which contains RSSI signal information about
   a message.
 - `Message` proto message has new field `Signal` of type `Signal`. Since its
   an object type, filters can test whether the field is set with `has()`
 - `State` proto message has new field `LastSignal` of type `Signal`. It contains
   the `Signal` value of the most recent message
 - `AircraftMap` now includes RSSI in the aircraft.json
 - `State` and `Message` are updated with new fields: `IndicatedAirSpeed`, `TrueAirSpeed`,
   `Mach`, `Roll`, `NavHeading`, `NavRoll`, `ADSBVersion`, `NACP`, `NACV`, `NICBaro`,
   `SIL`, `SILType`

### Changed

 - Fixes BEAST message processing - call `TrackUpdateFromMessage` ASAP so our message
   is updated with the processed information.
 - Fixes deadlock in tracker/map.go:updateJSON. Release projMu ASAP since ProjectHistory
   also needs it. Debugged using go-deadlock.
 - BEAST Server configuration: default to 30005 if nothing provided

## [0.0.5] - 2020-12-17

### Changed

 - `#62`: Fixed beast_producer so it reconnects if the network drops

pkg/config/config.go
package config

import (
	"fmt"
	"github.com/pkg/errors"
	"gopkg.in/yaml.v2"
	"io"
	"net/url"
	"os"
	"time"
)

const (
	// MailDriverSMTP - the only support SMTP driver
	MailDriverSMTP = "smtp"
	// DatabaseDriverMySQL - name of the mysql driver
	DatabaseDriverMySQL = "mysql"
	// DatabaseDriverPostgresql - name of the postgres driver
	DatabaseDriverPostgresql = "postgres"
	// DatabaseDriverSqlite3 - name of the sqlite3 driver
	DatabaseDriverSqlite3 = "sqlite3"
)

type (
	// Airports contains configuration for airport geolocation
	Airports struct {
		// DisableBuiltInAirports will prevent airtrack from loading
		// compiled in airport files if set to `true`
		DisableBuiltInAirports bool `yaml:"disable_builtin_airports"`
		// OpenAIPDirectories defined here will be scanned for .aip files
		OpenAIPDirectories []string `yaml:"openaip"`
		// CupDirecories defined here will be scanned for .cup files
		CupDirectories []string `yaml:"cup"`
	}

	// SMTPSettings - contains connection information for SMTP server.
	SMTPSettings struct {
		// Username - the SMTP username
		Username string `yaml:"username"`
		// Password - the SMTP password
		Password string `yaml:"password"`
		// Sender - the originating email address
		Sender string `yaml:"sender"`
		// Host - the SMTP server's hostname/ip
		Host string `yaml:"host"`
		// Port - the SMTP server port
		Port int `yaml:"port"`
		// TLS - whether to connect using TLS (port 587)
		TLS bool `yaml:"tls"`
		// MandatoryStartTLS - if set to true, connections to servers
		// which do not advertise STARTTLS support will cause an error.
		MandatoryStartTLS bool `yaml:"mandatory_starttls"`
		// NoStartTLS set to true disables opportunistic STARTTLS behaviour,
		// where the connection will be completely plaintext
		NoStartTLS bool `yaml:"nostarttls"`
	}

	// MapSettings contains configuration for providing
	// aircraft maps
	MapSettings struct {
		// Toggles whether map is enabled (default FALSE)
		Disabled bool `yaml:"disabled"`
		// HistoryInterval - number of seconds between new history files
		// (default: 30 seconds)
		HistoryInterval int64 `yaml:"history_interval"`
		// HistoryCount worth of history files will be kept. (default: 60)
		HistoryCount int `yaml:"history_count"`
		// Map interfaces to expose (default: dump1090 + tar1090)
		Services []string `yaml:"services"`
		// Interface webserver should listen on.
		Interface string `yaml:"interface"`
		// Port webserver should listen on (default: 8080)
		Port uint16 `yaml:"port"`
	}

	// EmailSettings is where email support is configured
	EmailSettings struct {
		// Driver - currently only 'smtp' is supported
		Driver string `yaml:"driver"`
		// SMTP points to a SMTPSettings struct for use with
		// the 'smtp' driver
		SMTP *SMTPSettings `yaml:"smtp"`
	}

	// Notifications - contains configuration of events to
	// send to user
	Notifications struct {
		// Email - destination for email events
		Email string `yaml:"email"`
		// Enabled - list of subscribed email events
		Enabled []string `yaml:"events"`
	}

	// ProjectMapSettings contains project level configuration
	// for the HTTP map UI
	ProjectMapSettings struct {
		// Disabled controls whether this project's map shall be
		// displayed on the HTTP map server. (default: false)
		Disabled bool `yaml:"disabled"`
	}

	// Project contains configuration for a single project
	Project struct {
		// Name - the name of the project (required)
		Name string `yaml:"name"`
		// Disabled controls whether the project should be running or not
		// this session. (default: false)
		Disabled bool `yaml:"disabled"`
		// Filter - an optional filter to apply to incoming messages
		Filter string
		// Map contains project level configuration for the map UI
		Map *ProjectMapSettings `yaml:"map"`
		// Notifications - per project configuration of event notifications
		Notifications *Notifications `yaml:"notifications"`
		// Features - per project extra features
		Features []string
		// ReopenSightings - whether to reopen a previously closed sighting
		// if a new sighting is within a certain timeframe
		ReopenSightings bool `yaml:"reopen_sightings"`
		// ReopenSightingsInterval - How long after an aircraft goes out of range
		// before we no longer reopen a recently closed session. Default 5m.
		ReopenSightingsInterval int `yaml:"reopen_sightings_interval"`
		// OnGroundUpdateThreshold - how many on_ground messages before we propagate
		// the change in status
		OnGroundUpdateThreshold *int64 `yaml:"onground_update_threshold"`
		// LocationUpdateInterval enforces a minimum duration between location updates.
		// When set to zero, all location messages are accepted.
		// Units are seconds.
		LocationUpdateInterval *int64 `yaml:"location_update_interval"`
	}

	// Database - connection information about the database
	Database struct {
		// Driver to use for connections: sqlite3, mysql, postgresql
		Driver string `yaml:"driver"`
		// Host applies to mysql/postgresql - the DB server to connect to
		Host string `yaml:"host"`
		// Port applies to mysql/postgresql - the DB server port to connect to
		Port int `yaml:"port"`
		// Username is the username used when connecting to mysql/postgresql
		Username string `yaml:"username"`
		// Password is the password used when connecting to mysql/postgresql
		Password string `yaml:"password"`
		// Database - filesystem path to sqlite3 file, or database name on mysql/postgresql
		Database string `yaml:"database"`
	}

	// Metrics - contains configuration for prometheus metrics
	Metrics struct {
		// Enabled - control whether to enable metrics. Default to off.
		Enabled bool `yaml:"enabled"`
		// Interface - interface to bind on. If empty, default is "0.0.0.0"
		Interface string `yaml:"interface"`
		// Port - port to listen for metrics HTTP server. If empty, default is 8080.
		Port int `yaml:"port"`
	}

	// AdsbxConfig contains configuration for the ADSB Exchange data source
	AdsbxConfig struct {
		// Custom ADSB Exchange URL (not required, but useful if
		// you've a proxy setup)
		APIURL string `yaml:"url"`
		// ADSB Exchange API key
		APIKey string `yaml:"apikey"`
	}

	// BeastConfig contains configuration for a single BEAST server
	BeastConfig struct {
		// Name for this beast server
		Name string `yaml:"name"`
		// IP or hostname for beast server
		Host string `yaml:"host"`
		// Port for beast services (Optional, defaults to 30005)
		Port *uint16 `yaml:"port"`
	}

	// Config - represents the yaml block in the main config file.
	Config struct {
		// TimeZone - optional timezone to override system default
		TimeZone *string `yaml:"timezone"`
		// AdsbxConfig - optional element, set if using ADSBExchange API
		AdsbxConfig *AdsbxConfig `yaml:"adsbx"`
		// Beast - list of beast server configs
		Beast []BeastConfig `yaml:"beast"`
		// Airports - where directories of airport location files are configured
		Airports *Airports `yaml:"airports"`
		// EmailSettings - configuration of email driver.
		EmailSettings *EmailSettings `yaml:"email"`
		// Database - configuration of the database driver.
		Database Database `yaml:"database"`
		// Metrics - configuration of prometheus metrics
		Metrics *Metrics `yaml:"metrics"`
		// MapSettings - configuration of the HTTP map server
		MapSettings *MapSettings `yaml:"map"`
		// Sighting - some global defaults for sighting configuration
		Sighting struct {
			Timeout *int64 `yaml:"timeout"`
			// LocationUpdateInterval sets a default LocationUpdateInterval
			// to be used by projects which don't specify
			LocationUpdateInterval *int64 `yaml:"location_update_interval"`
		} `yaml:"sighting"`
		// Projects - list of project configurations
		Projects []Project `yaml:"projects"`
	}

	// ProjectsConfig - represents the yaml block in a projects configuration file
	ProjectsConfig struct {
		// Projects - list of project configurations
		Projects []Project `yaml:"projects"`
	}
)

// GetTimeLocation returns a *time.Location (or an error on failure)
// that is specified in the config file, or the system default if none
// is provided.
func (c *Config) GetTimeLocation() (*time.Location, error) {
	if c.TimeZone == nil {
		return time.Now().Location(), nil
	}

	tz := *c.TimeZone
	var err error
	loc, err := time.LoadLocation(tz)
	if err != nil {
		return nil, err
	}
	return loc, nil
}

// DataSource returns a connection string for the configured database driver
func (d *Database) DataSource(loc *time.Location) (string, error) {
	switch d.Driver {
	case "":
		return "", errors.New("no database driver configured")
	case DatabaseDriverMySQL:
		return fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?parseTime=true&loc=%s",
			d.Username, d.Password, d.Host, d.Port, d.Database, url.PathEscape(loc.String())), nil
	case DatabaseDriverPostgresql:
		return fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s",
			d.Host, d.Port, d.Username, d.Password, d.Database), nil
	case DatabaseDriverSqlite3:
		return fmt.Sprintf("file:%s?parseTime=true&loc=%s",
			d.Database, url.PathEscape(loc.String())), nil
	default:
		return "", errors.Errorf("unsupported database driver `%s`", d.Driver)
	}
}

// ReadConfigFromFile will read `filepath` and attempt to parse into
// a Config structure. An error will be returned if duplicated project
// names are encountered.
func ReadConfigFromFile(filepath string) (*Config, error) {
	if filepath == "" {
		return nil, errors.New("empty configuration file path provided")
	}
	f, err := os.Open(filepath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	return ReadConfig(f)
}

// ReadConfig will decode the provided reader into a Config structure.
// An error will be returned if duplicated project names are encountered.
func ReadConfig(r io.Reader) (*Config, error) {
	cfg := Config{}
	decoder := yaml.NewDecoder(r)
	err := decoder.Decode(&cfg)
	if err != nil {
		return &cfg, err
	}
	projMap := make(map[string]struct{})
	for _, proj := range cfg.Projects {
		_, ok := projMap[proj.Name]
		if ok {
			return nil, errors.Errorf("duplicated project name: %s", proj.Name)
		}
	}
	return &cfg, nil
}

// ReadProjectsConfigFromFile will read `filepath` and attempt to parse into
// a Config structure. An error will be returned if duplicated project
// names are encountered.
func ReadProjectsConfigFromFile(filepath string) (*ProjectsConfig, error) {
	if filepath == "" {
		return nil, errors.New("empty project file path provided")
	}
	f, err := os.Open(filepath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	return ReadProjectsConfig(f)
}

// ReadProjectsConfig will decode the provided reader into a ProjectsConfig structure.
// An error will be returned if duplicated project names are encountered.
func ReadProjectsConfig(r io.Reader) (*ProjectsConfig, error) {
	cfg := ProjectsConfig{}
	decoder := yaml.NewDecoder(r)
	err := decoder.Decode(&cfg)
	if err != nil {
		return &cfg, err
	}
	projMap := make(map[string]struct{})
	for _, proj := range cfg.Projects {
		_, ok := projMap[proj.Name]
		if ok {
			return nil, errors.Errorf("duplicated project name: %s", proj.Name)
		}
	}
	return &cfg, nil
}

// ReadConfigs will decode the provided 'main' configFile, along with
// any extra project only files, and return the initialized configuration.
// An error will be returned if duplicated project names are encountered.
func ReadConfigs(configFile string, projectsFiles []string) (*Config, error) {
	cfg, err := ReadConfigFromFile(configFile)
	if err != nil {
		return nil, errors.Wrap(err, "reading main config file")
	}
	// build up current projects. ReadConfigFromFile guarantees no duplicates.
	projects := make(map[string]struct{})
	for _, proj := range cfg.Projects {
		projects[proj.Name] = struct{}{}
	}
	for _, projectsFile := range projectsFiles {
		projConfig, err := ReadProjectsConfigFromFile(projectsFile)
		if err != nil {
			return nil, errors.Wrapf(err, "failed to read projects config file %s", projectsFile)
		}
		for _, proj := range projConfig.Projects {
			// ensure no new duplicates
			if _, ok := projects[proj.Name]; ok {
				return nil, errors.Wrapf(err, "duplicated project name: %s", proj.Name)
			}
			cfg.Projects = append(cfg.Projects, proj)
		}
	}
	return cfg, nil
}
